docs/src/content/docs/book/security-best-practices.mdx

```diff
@@ -0,0 +1,435 @@
+---
+title: Security Best Practices
+---
+
+[//]: # (✅❌)
+
+In Tact smart contracts there are various anti-patterns and potential attacks developers need to be cautious of. These can impact the security, efficiency, and correctness of the contracts. Below, are covered both common anti-patterns and attacks specific to Tact smart contracts.
+
+## Send sensitive data on chain
+
+The whole smart contract computation is transparent and if you had some confidential values in runtime these could be retrieved with a simple emulation.
+
+##### Do's ✅
+
+Do not send sensitive data on chain.
+
+##### Don'ts ❌
+
+```tact
+import "@stdlib/deploy";
+
+message Login {
+    privateKey: Int as uint256;
+    signature: Slice;
+    data: Slice;
+}
+
+contract Test with Deployable {
+    receive(msg: Login) {
+        let publicKey = getPublicKey(msg.privateKey);
+
+        require(checkDataSignature(msg.data, msg.signature, publicKey), "Invalid signature!");
+    }
+}
+```
+
+
+## Misuse of signed integers
+
+Unsigned integers are safer because they prevent most errors by design, while signed integers can lead to unpredictable consequences if not used carefully. Therefore, signed integers should only be used when absolutely necessary.
+
+##### Do's ✅
+
+Prefer using unsigned integers unless signed integers are specifically required.
+
+##### Don'ts ❌
+
+The following is an example of incorrect use of a signed integer. In the `Vote` message, the `votes` type is `int32`. This can lead to fraud when the attacker sends the negative number of votes instead of positive one.
+
+```tact
+message Vote {
+    votes: Int as int32;
+}
+
+contract Sample {
+    votes: Int as uint32 = 0;
+
+    receive(msg: Vote) {
+        self.votes += msg.votes;
+    }
+}
+```
+
+## Invalid throw values
+
+Exit codes $0$ and $1$ indicate normal TVM execution of the compute phase of the transaction. Execution can be unexpectedly aborted by calling [`throw(){:tact}`](/ref/core-debug#throw) (or [similar functions](/ref/core-debug)) directly with exit codes $0$ and $1$, which can make debugging very difficult since this aborted execution would be considered normal.
+
+##### Do's ✅
+
+Prefer using `require` to validate some conditions.
+
+```tact
+require(isDataValid(msg.data), "Invalid data!");
+```
+
+##### Don'ts ❌
+
+Don't throw $0$ or $1$ directly.
+
+```tact
+throw(0);
+throw(1);
+```
+
+## Insecure random numbers
+
+Note that [`random()`](/ref/core-random#random) function is pseudo-random and depends on [logical time](https://docs.ton.org/develop/smart-contracts/guidelines/message-delivery-guarantees#what-is-a-logical-time). A hacker can predict random by brute-forcing the logical time in the current block.
```

Comment:
Not really. If a hacker does not own a Validator Node (or does not agreed with validator ) he can do nothing if you yse randomize_lt() before generating a random.

Reply:
But no one should use random in external_message receivers. Random there can be hacked even if person is not a Validator and even if you use randomize_lt()

Reply:
However, a commit-and-disclose scheme is sometimes possible. First, both sides generate offchain a number each, and send a hash of their numbers to the contract. After both hashes have arrived, the original numbers can be sent - and their sum can be used, for example, as a randomizer.